### PR TITLE
file_system: bypass file allocate implementation in fs2

### LINKED ISSUE #10688: tikv panic with "no such file or directory" when disk space is not enough
### What version of TiKV are you using?
```
[root@centos76_vm tikv-20161]# bin/tikv-server -V
TiKV
Release Version:   5.1.0-alpha
Edition:           Community
Git Commit Hash:   353897753d27fedc6343a0b84bcab454cdec10d3
Git Commit Branch: master
UTC Build Time:    2021-07-27 09:39:00
Rust Version:      rustc 1.53.0-nightly (16bf626a3 2021-04-14)
Enable Features:   jemalloc mem-profiling portable sse protobuf-codec test-engines-rocksdb cloud-aws cloud-gcp
Profile:           dist_release
```

### Steps to reproduce

1. disk full and tikv restart repeatly with error log
```
[2021/08/09 12:05:06.956 +08:00] [FATAL] [server.rs:1177] ["failed to create raft engine: Storage Engine IO error: No space left on deviceWhile appending to file: /tidb/eindex/data/tikv-20161/       raft/226776.sst: No space left on device"]
```
2. remove the space_hodler_file manually

### What did you expect?

Don't know. Maybe restart with some error message, instead of panic with "No such file or directory".

### What did happened?
1. tikv panic with following log. 
```
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }"] [backtrace="stack backtrace:
   0: tikv_util::set_panic_hook::{{closure}}
             at components/tikv_util/src/lib.rs:462
   1: std::panicking::rust_panic_with_hook
             at library/std/src/panicking.rs:595
   2: std::panicking::begin_panic_handler::{{closure}}
             at library/std/src/panicking.rs:497
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at library/std/src/sys_common/backtrace.rs:141
   4: rust_begin_unwind
             at library/std/src/panicking.rs:493
   5: core::panicking::panic_fmt
             at library/core/src/panicking.rs:92
   6: core::result::unwrap_failed
             at library/core/src/result.rs:1355
   7: core::result::Result<T,E>::unwrap
             at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/result.rs:1037
      server::server::TiKVServer<ER>::init_fs
             at components/server/src/server.rs:416
      server::server::run_tikv
             at components/server/src/server.rs:151
   8: tikv_server::main
             at cmd/tikv-server/src/main.rs:181
   9: core::ops::function::FnOnce::call_once
             at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/ops/function.rs:227
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/std/src/sys_common/backtrace.rs:125
  10: main
  11: __libc_start_main
  12: <unknown>
```
2. I checked the `space_holder_file` file size, and it is 0, which seems also strange.
```
[root@centos76_vm tikv-20161]# du -hs /tidb/eindex/data/tikv-20161/space_placeholder_file
0	/tidb/eindex/data/tikv-20161/space_placeholder_file
```

### PR DESCRIPTION
**Pull request by tabokie**
Signed-off-by: tabokie <xy.tao@outlook.com>

### What is changed and how it works?

Close #10688

Patch fs2 to https://github.com/tabokie/fs2-rs/tree/tikv. In which, `posix_fallocate` is replaced with `fallocate` (https://github.com/danburkert/fs2-rs/pull/42).

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests
- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/file_system/src/file.rs [hunk 1]
```diff
@@ -159,7 +159,95 @@ impl File {
     }
 
     pub fn allocate(&self, len: u64) -> io::Result<()> {
-        self.inner.allocate(len)
+        // Bypass fs2::FileExt::allocate because of its incorrect error handling.
```
Comment (Connor1996) [inline components/file_system/src/file.rs:162]: wait the upstream to merge it?

Comment (tabokie) [reply components/file_system/src/file.rs:162]: upstream has no activity for 4 years now, I'm feeling negative about it. Plus we can always revert the change if upstream merges it afterwards.

Comment (Connor1996) [reply components/file_system/src/file.rs:162]: maybe use our fork?

Comment (tabokie) [reply components/file_system/src/file.rs:162]: I think that would be an overkill, we don’t have other changes in the future.

Comment (Connor1996) [reply components/file_system/src/file.rs:162]: use your branch is enough, no need to push to tikv repo

Comment (tabokie) [reply components/file_system/src/file.rs:162]: Oh, make sense.

**Review by Connor1996 — APPROVED** [review]
LGTM

**Review by hunterlxt — APPROVED** [review]
LGTM

**Pull request merged**